edgelet/aziot-edged/src/main.rs

```diff
@@ -145,15 +162,27 @@ async fn run() -> Result<(), EdgedError> {
     // Set signal handlers for SIGTERM and SIGINT.
     set_signal_handlers(watchdog_tx);
 
-    // Run aziot-edged until the shutdown signal is received. This also runs the watchdog periodically.
-    let shutdown_reason = watchdog::run_until_shutdown(
-        settings,
+    let watchdog = watchdog::run_until_shutdown(
+        settings.clone(),
         &device_info,
-        runtime,
+        runtime.clone(),
         &identity_client,
         watchdog_rx,
-    )
-    .await?;
+    );
+    let image_gc = image_gc::image_garbage_collect(settings.clone(), &runtime, image_use_data);
+
+    let shutdown_reason: WatchdogAction;
+    tokio::select! {
+        watchdog_finished = watchdog => {
+            log::info!("watchdog finished");
+            shutdown_reason = watchdog_finished?;
+        },
+        image_gc_finished = image_gc => {
```

Comment:
Hmm... I think the only time we exit GC is if there's an invalid config OR if state file creation fails at startup. This is because the feature is enabled by default and not exiting might make the user mistakenly think that we are pruning images when we aren't. In this case, logging and error and continuing with MIGC in a disabled state wouldn't be the right thing to do IMO. We should fail loudly.

Apart from that though, we never exit GC. We just log and move on.

Even in the case where migc is explicitly disabled, we return a future that never completes.


I'm uncertain if more is needed here, but I'm open to the possibility.